### ISSUE TITLE
BPA Rule "Avoid Division" in error in TE 3.16.1

### ISSUE THREAD

**Issue by edhans** [TabularEditor/TabularEditor3#1323]
### Description

The Avoid Division BPA rule is showing red in all of my models. 
![image](https://github.com/TabularEditor/TabularEditor3/assets/45298358/7dfb88f9-4d20-40e6-b1cf-f5ab1075d670)

The full error says "No property or field 'Type' exists in 'Int32'

I haven't updated my BPA rule file since April 24, 2024. The JSON for this rule is below, and I've never touched it to my knowledge

  ```
{
    "ID": "DAX_DIVISION_COLUMNS",
    "Name": "Avoid division (use DIVIDE function instead)",
    "Category": "DAX Expressions",
    "Description": "Calculated Columns, Measures or Calculated Tables should not use the division symbol in their expressions (/) unless the denominator is a constant value. Instead, it is advised to always use the DIVIDE(<numerator>,<denominator>) function.",
    "Severity": 3,
    "Scope": "Measure, CalculatedColumn, CalculatedTable, CalculationItem",
    "Expression": "Tokenize().Any(\n    Type = DIV and\n    Next.Type <> INTEGER_LITERAL and\n    Next.Type <> REAL_LITERAL\n)",
    "CompatibilityLevel": 1200
  },
```

I have a few measures where I am using the divide operator for speed where I know it is impossible to have an invalid denominator, but those aren't showing. The rule is just red.

I don't know if this is new in 3.16.1 or I just didn't notice it in 3.16 over the last 2-3 days.

### Tabular Editor 3 Version

3.16.1

### Windows Version

11

**Comment by otykier** [comment]
Thanks Ed - looking into it (the BPA rule works in TE2, so it should definitely work in TE3 as well)

**Comment by edhans** [comment]
Thanks. It is very likely I have a measure or two in this model that would trigger this rule, but I've told it to ignore them, so it could be that combination of things causing it to show up in red like this. And I did just install 3.16.2 and it is still there. I know the .2 update had nothing to do with this, but just letting you know I am on the very latest.

**Comment by edhans** [comment]
This is just broken. I created a test model (other reasons) and it only has one COUNTROWS() measure. this BPA rule is showing up in red with the same error.

**Comment by mlonsk** [comment]
Hi Ed
We have confirmed this bug and reproduced it on our end. 
We will let you know once it is fixed!

Until then, you can use the following from Microsoft's BPA rules instead.
Your Tokonize version is, of course, an improvement, but something is better than nothing I would venture. 

```
  {
    "ID": "USE_THE_DIVIDE_FUNCTION_FOR_DIVISION",
    "Name": "[DAX Expressions] Use the DIVIDE function for division",
    "Category": "DAX Expressions",
    "Description": "Use the DIVIDE  function instead of using \"/\". The DIVIDE function resolves divide-by-zero cases. As such, it is recommended to use to avoid errors.\r\n\r\nReference: https://docs.microsoft.com/power-bi/guidance/dax-divide-function-operator",
    "Severity": 2,
    "Scope": "Measure, CalculatedColumn, CalculationItem",
    "Expression": "RegEx.IsMatch(Expression,\"\\]\\s*\\/(?!\\/)(?!\\*)\")\r\nor\r\nRegEx.IsMatch(Expression,\"\\)\\s*\\/(?!\\/)(?!\\*)\")",
    "CompatibilityLevel": 1200
  },
```

**Comment by otykier** [comment]
Hi @edhans 
Happy to let you know that this has also been fixed in [3.17.0](https://docs.tabulareditor.com/te3/other/release-notes/3_17_0.html?tabs=downloads).

The problem was that your BPA rule used the `Next` property within the context of DAX tokens returned by the `Tokenize()` function. This property is used to access the next token after the current one. Unfortunately, a [`NEXT()`](https://dax.guide/next/) function was recently added to the DAX language (it's used for visual calculations), and since the Dynamic LINQ expression language of BPA rules is case-**insensitive**, the original BPA expression now became invalid because of this ambiguity.

The way we fixed this, is to resolve ambiguities in rule expressions by performing a case-**sensitive** match first. In the context of DAX tokens returned by `Tokenize()`, this now means that uppercase `NEXT` refers to the `NEXT()` function in DAX, while propercase `Next` refers to the next token after the current one. In general, all functions and DAX keywords are uppercase, but `NEXT` is the only one currently that has an ambiguity with the properties on the tokens returned by `Tokenize()`. For reference, these are:

```csharp
// Gets the next non-comment, non-whitespace token after the current token:
public DaxToken Next { get; }

// Gets the previous non-comment, non-whitespace token after the current token:
public DaxToken Prev { get; }

// Gets the integer constant that represents the type of the current token. This matches uppercase tokens defined on the DaxToken class, such as DaxToken.CALCULATE, DaxToken.EVALUATE, DaxToken.STRING_LITERAL, etc.
public int Type { get; }

// Gets the integer position of the current token within the DAX expression:
public int TokenIndex { get; }

// Gets the character position where the current token starts within the DAX expression:
public int StartIndex { get; }

// Gets the character position where the current token ends within the DAX expression:
public int StopIndex { get; }

// true if the current token is a comment or whitespace. false otherwise.
public bool CommentOrWhitespace { get; }

// The substring of the original DAX expression, that corresponds to the current token:
public string Text { get; }
